stellargraph/data/explorer.py

```diff
@@ -789,3 +793,136 @@ def _check_neighbourhood_sizes(self, in_size, out_size):
+                )
+                walks.append(walk)
+        return walks
+
+    def _exp_biases(self, times, t_0, decay):
+        # t_0 assumed to be smaller than all time values
+        dist = np.exp(t_0 - np.array(times) if decay else np.array(times) - t_0)
+        sum_dist = np.sum(dist)
+        return dist / sum_dist
+
+    def _biases(self, times, t_0, bias_type, is_forward):
+        if bias_type == "exponential":
+            # exponential decay bias needs to be reversed if looking backwards in time
+            return self._exp_biases(times, t_0, decay=is_forward)
+        else:
+            raise ValueError("Unsupported bias type")
+
+    def _step(self, node, time, is_forward, bias_type, np_rs):
+        """
+        Perform 1 temporal step from a node. Returns None if a dead-end is reached.
+
+        """
+
+        def check_time(t):
+            # time is None indicates no temporal filtering (i.e. initial edge selection)
+            return (
+                (time is None)
+                or (is_forward and t > time)
+                or (not is_forward and t < time)
+            )
+
+        neighbours = [
+            (neighbour, t)
+            for neighbour, t in self.graph.neighbors(node, include_edge_weight=True)
+            if check_time(t)
+        ]
+
+        if len(neighbours) > 0:
+            if bias_type is not None:
+                times = [t for _, t in neighbours]
+
+                # time is None indicates we should obtain the minimum available time for t_0
+                t_0 = time if time is not None else min(times)
+                biases = self._biases(
+                    times, t_0, bias_type=bias_type, is_forward=is_forward
+                )
+            else:
+                # default to uniform random sampling
+                biases = None
+            chosen_index = np_rs.choice(len(neighbours), p=biases)
+            next_node, next_time = neighbours[chosen_index]
+            return next_node, next_time
+        else:
+            return None
+
+    def _walk(self, node, length, bidirectional, initial_edge_bias, walk_bias, np_rs):
+        def step(n, t, is_forward, bias_type=walk_bias):
+            return self._step(
+                n, time=t, is_forward=is_forward, bias_type=bias_type, np_rs=np_rs
+            )
+
+        # sample 1 edge to obtain starting time - any biases are applied as though we are
+        # looking backwards in time.
+        forward = step(node, None, is_forward=False, bias_type=initial_edge_bias)
+        if forward is None:
+            return [node]
+        backward = node, forward[1]
+
+        walk = []
+
+        # take steps until walk is of correct length or reached dead ends
+        while len(walk) < length:
+            if forward is None and backward is None:
+                break
+            if forward is not None:
+                n, t = forward
+                walk.append(n)
```

Comment:
I think, if `bidirectional` is False, `node` never ends up in the walk: the only modification is this, which is appending the next node chosen by `step`, not `node` itself.

Reply:
If `bidirectional` is false, then the first iteration of the while loop has `backward = node` (line 911), so node does get added, then `backward` is set to `None`. I'd be interested in suggestions to achieve the same thing with more clarity.. I found the whole bidirectional logic in the while loop a bit clunky to write, I ended up going with something shorter rather than spelling out the 0th iteration first then doing a loop if that makes sense.

Reply:
Huh, that's subtle and I can definitely see where you're coming from with the clunkiness.

I don't love it, but how about something like:

```python
walk = deque([node])

forward_time = None
backward_time = None

while len(walk) < length:
    if len(walk) == 1:
        # first step is special
        forward, forward_time = step(walk[0], None, is_forward=False, bias_type=initial_edge_bias)
        walk.append(forward)
        if is_bidirectional:
            backward_time = forward_time
    elif forward_time is not None:
        forward, forward_time = step(walk[-1], forward_time, is_forward=True)
        walk.append(forward)

    if len(walk) == length:
        break

    if backward_time is not None:
        backward, backward_time = step(walk[0], backward_time, is_forward=False)
        walk.appendleft(backward)

    if backward_time is None and forward_time is None:
        break
```

With this model, the `walk` indexing and mutation logic could potentially be folded into `step`:

```python
walk = collections.deque([node])

def step(t, is_forward, is_first=False):
    n = walk[-1] if is_forward else walk[0]
    bias = initial_bias_type if is_first else bias_type
    result = self._step(n, t, ...)

    if result is None:
        return None

    next_node, next_time = result
    # the first step treats times as backwards, but still appends as if forward
    if is_forward or is_first: 
        walk.append(next_node)
    else:
        walk.prepend(next_node)

    return next_time
```

and the above becomes:

```python
forward_time = None
backward_time = None

while len(walk) < length:
    if len(walk) == 1:
        # first step is special
       forward_time = step(None, is_forward=False, is_first=True)
       if is_bidirectional:
           backward_time = forward_time
    elif forward_time is not None:
        forward_time = step(forward_time, is_forward=True)

    if len(walk) == length:
        break

    if backward_time is not None:
        backward_time = step(backward_time, is_forward=False)

    if backward_time is None and forward_time is None:
        # no eligible neighbours
        break
```

Reply:
Thanks I'll use that 👍  - I think using `if` instead of `elif` would maintain original behaviour of doing a forward step first after the initial edge selection